File: ojscala/src/main/scala/com/zhourui/leetcode/lc0200_numberofisland.scala
package com.zhourui.leetcode


// 思路 
package lc0200_numberofisland {
  object Solution {
    def numIslands(grid: Array[Array[Char]]): Int = {
      if(grid.isEmpty || grid(0).isEmpty){
        return 0
      }
      val m = grid.size
      val n = grid(0).size;

      //val dp:Array[Array[Int]];//vector<vector<int>> dp(m,std::vector<int>(n));
      //val dp = Array.ofDim[Int](m,n)
      val dp = Array.fill(m,n)(0)

      var count=0;
      for (i<-0 until m) {
        for (j<-0 until n) {
          if (dp(i)(j)==0) {
            if (bfs(grid,dp,(i,j)))
              count +=1;
          }
        }
      }
      count;
    }

    def bfs(grid:Array[Array[Char]], dp:Array[Array[Int]],pos:(Int,Int)):Boolean = {
      val m = grid.length
      val n = grid(0).length
      if (pos._1>=m || pos._2>=n || pos._1<0 || pos._2<0) {
        return false
      }
      if (dp(pos._1)(pos._2) == 1) {
        false
      } else if (dp(pos._1)(pos._2) == 0 && grid(pos._1)(pos._2) == '1') {
        dp(pos._1)(pos._2) = 1
        // right
        bfs(grid, dp, (pos._1, pos._2 + 1))
        // down
        bfs(grid, dp, (pos._1 + 1, pos._2))
        // up
        bfs(grid, dp, (pos._1 - 1, pos._2))
        // left
        bfs(grid, dp, (pos._1, pos._2 - 1))
        true;
      } else {
        false
      }
    }
  }
}

File: ojjava/src/main/java/com/zhourui/leetcode/Lc0200_numberofisland.java
package com.zhourui.leetcode;

import com.zhourui.codech.BaseSolution;

// 这种matrix的，多半和BFS,DFS搜索有关
//可以使用DIR =[1,1,-1
public class Lc0200_numberofisland extends BaseSolution {
    class Solution {
        int[] DIR = new int[]{0, 1, 0, -1, 0};  // 4 direction
        public int numIslands(char[][] grid) {
            if (grid.length == 0) return 0;
            int m=grid.length,n = grid[0].length;

            boolean[][] visited = new boolean[m][n];
            int ans=0;
            for (int r=0;r<m;r++) {
                for (int c=0;c<n;c++) {
                    if (!visited[r][c]) {
                        ans += helper(m,n,grid,visited,r,c);
                    }
                }
            }
            return ans;
        }

        int helper(int m, int n, char[][]grid, boolean[][] visited, int r, int c) {
            if (r<0 || r>m || c<0 || c>n || grid[r][c]=='0' || visited[r][c]) {
                return 0;
            }
            visited[r][c] = true;
            for (int i=0;i<4;i++) {
                helper(m,n,grid,visited,r+DIR[i],c+DIR[i+1]);
            }
            return 1;
        }
    }
}
